procrastinate/sql/migrations/02.06.00_01_add_cancel_states.sql

```diff
@@ -0,0 +1,115 @@
+ALTER TYPE procrastinate_job_status ADD VALUE 'cancelled';
+ALTER TYPE procrastinate_job_status ADD VALUE 'aborting';
+ALTER TYPE procrastinate_job_status ADD VALUE 'aborted';
+
+ALTER TYPE procrastinate_job_event_type ADD VALUE 'request_to_abort' BEFORE 'scheduled';
+ALTER TYPE procrastinate_job_event_type ADD VALUE 'aborted' BEFORE 'scheduled';
+
+CREATE FUNCTION procrastinate_cancel_job(job_id bigint, abort boolean, delete_job boolean)
+    RETURNS bigint
+    LANGUAGE plpgsql
+AS $$
+DECLARE
+    _job_id bigint;
+BEGIN
+    IF delete_job THEN
+        DELETE FROM procrastinate_jobs
+        WHERE id = job_id AND status = 'todo'
+        RETURNING id INTO _job_id;
+    END IF;
+    IF _job_id IS NULL THEN
+        IF abort THEN
+            UPDATE procrastinate_jobs
+            SET status = CASE
+                WHEN status = 'todo' THEN 'cancelled'::procrastinate_job_status
+                WHEN status = 'doing' THEN 'aborting'::procrastinate_job_status
+            END
+            WHERE id = job_id AND status IN ('todo', 'doing')
+            RETURNING id INTO _job_id;
+        ELSE
+            UPDATE procrastinate_jobs
+            SET status = 'cancelled'::procrastinate_job_status
+            WHERE id = job_id AND status = 'todo'
+            RETURNING id INTO _job_id;
+        END IF;
+    END IF;
+    RETURN _job_id;
+END;
+$$;
+
+CREATE OR REPLACE FUNCTION procrastinate_finish_job(job_id bigint, end_status procrastinate_job_status, delete_job boolean)
+    RETURNS void
+    LANGUAGE plpgsql
+AS $$
+DECLARE
+    _job_id bigint;
+BEGIN
+    IF end_status NOT IN ('succeeded', 'failed', 'aborted') THEN
+        RAISE 'End status should be either "succeeded", "failed" or "aborted" (job id: %)', job_id;
+    END IF;
+    IF delete_job THEN
+        DELETE FROM procrastinate_jobs
+        WHERE id = job_id AND status IN ('todo', 'doing', 'aborting')
+        RETURNING id INTO _job_id;
+    ELSE
+        UPDATE procrastinate_jobs
+        SET status = end_status,
+            attempts =
+                CASE
+                    WHEN status = 'doing' THEN attempts + 1
+                    ELSE attempts
+                END
+        WHERE id = job_id AND status IN ('todo', 'doing', 'aborting')
+        RETURNING id INTO _job_id;
+    END IF;
+    IF _job_id IS NULL THEN
+        RAISE 'Job was not found or not in "doing", "todo" or "aborting" status (job id: %)', job_id;
+    END IF;
+END;
+$$;
+
+CREATE OR REPLACE FUNCTION procrastinate_trigger_status_events_procedure_update()
```

Comment:
And so the circle closes 😉. Does this mean I have to "create or replace" the `procrastinate_trigger_status_events_update` trigger (because this is the one that executes this procedure)?

Reply:
I guess so. If you can do a quick manual check by applying the previous schema, then the migration, then have the trigger execute and check that it's the right code, I'll be convinved, no need for an automated test or anything :)

Reply:
The [PostgreSQL documentation](https://www.postgresql.org/docs/current/sql-createfunction.html) also says:

> Use CREATE OR REPLACE FUNCTION to change a function definition without breaking objects that refer to the function

And this is what we do here. But I can double-check and try it out.

Reply:
Ok, I tried it out. It still seems to work without recreating the trigger. Here is what I did:

1. Started with an empty db
2. Checked out the main branch
3. Applied the schema
4. Ran the async demo with one job
5. Checked that the events were triggered (in the db table)
6. Checked out the cancel-job-feature branch
7. Applied the add_cancel_states migration
8. Ran the async demo again with one job
9. Checked that events were triggered again

Everything seemed to work fine. The resulting table:

```
procrastinate=# SELECT * FROM procrastinate_events;
 id | job_id |   type    |              at               
----+--------+-----------+-------------------------------
  1 |      1 | deferred  | 2024-06-25 09:31:44.600697+00
  2 |      1 | started   | 2024-06-25 09:31:44.613338+00
  3 |      1 | succeeded | 2024-06-25 09:31:44.623551+00
  4 |      2 | deferred  | 2024-06-25 09:39:07.207435+00
  5 |      2 | started   | 2024-06-25 09:39:07.219847+00
  6 |      2 | succeeded | 2024-06-25 09:39:07.231104+00
(6 rows)
```

Reply:
Perfect.